### PR TITLE
Documentation and Tests for InitActions

### PR DESCRIPTION
**Pull request by nora-ambroz**
This change is [Reviewable](https://reviewable.io/reviews/aliz-ai/jarvis/17)

### CODE REVIEW

File: README.MD [hunk 1]
```diff
@@ -141,6 +141,63 @@ The TalendAPI context type requires an _apiUrl_ and an _apiKey_ parameter to con
...
+This order never changes and each phase starts only after the previous has finished.
+This part of the documentation details the role of each action and how the events can be customized during the steps.
+
+### Initiation Action
+The initiation action prepares the environment for the upcoming test execution. 
+During this phase the test runner can automatically create the prerequisites of the test.
+For example, it can load data or clean up after a previous run.
+If the initiation is correctly set up, then every condition will be provided to repeat the test.
+
+If the __testSuite.json__ allows autodetect, then Jarvis parses the configuration for the initiation actions from the folder structure of the test case.
```
Comment (nora-ambroz) [inline README.MD:157]: This seems vague. What else should be said about the testSuite.json here?

File: src/main/java/ai/aliz/jarvis/service/shared/platform/SFTPService.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            
+            String fileName = localFile.getFileName().toString();
+            
+            SFTPService.RemoteFileMeta remoteFileMeta = prepareRemoteFileMeta(fileName);
+            String remoteBasePath = sftpContext.getParameter(REMOTE_BASE_PATH);
+            String fileRemotePath = remoteBasePath + "/" +
+                    remoteFileMeta.getFilename();
+            
+            synchronized (sftpContext) {
+                // this is ugly, but a prettier solution would require a bigger redesign
```
Comment (nora-ambroz) [inline src/main/java/ai/aliz/jarvis/service/shared/platform/SFTPService.java:86]: What should I do here?

File: README.MD (hunk 1 above)
Comment (fpeter8) [reply README.MD:157]: - all possible fields (I found `src/main/resources/test_bqQuery/testSuite.json` with other fields, but that may not be all)
- this states what happens if we allow autodetect. What happens if we do not? Where does the configuration come from if not from the folders? The file itself? There is nothing about it in the documentation

File: src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    }
+    
+    @Override
+    public void executeScript(String query, Context context) {
+        List<String> deletes = Lists.newArrayList();
+        List<String> inserts = Lists.newArrayList();
+        Arrays.stream(query.split(";"))
+              .map(String::trim)
+              .filter(e -> !e.isEmpty())
+              .map(String::toUpperCase)
```
Comment (nora-ambroz) [inline src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java:80]: Should not change to upper case.
Fixed on init-action-integration-tests branch.

File: README.MD (hunk 1 above)
Comment (csabakassai) [reply README.MD:157]: If autodetect is not enabled then it is defined in the file itself.  
Also please document the config for different init actions separately
1. BQLoad
2. SQL script execution
3. SFTP load

File: README.MD [hunk 4]
```diff
@@ -141,6 +141,63 @@ The TalendAPI context type requires an _apiUrl_ and an _apiKey_ parameter to con
 
 Any other parameters not listed here can be used as substitution variables in files (when the key is used it is replaced by the value during query execution).
 
+---
+## Actions
```
Comment (csabakassai) [inline README.MD:145]: In the code you call this 'Step'. I think that is better name, please use also stepo here in the docs

File: src/test/java/ai/aliz/jarvis/util/TestJarvisUtil.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        assertEquals("select * from REPLACED_TABLE where 1=1;", JarvisUtil.resolvePlaceholders(query, parameters));
+    }
+    
+    @Test
+    public void testResolvePlaceholdersNotPresentParam() {
+        String query = "select * from {{table}} where {{column}}={{value}};";
+        Map<String, String> parameters = new HashMap<>();
+        parameters.put("table", "REPLACED_TABLE");
+        parameters.put("value", "1");
+        assertEquals("select * from REPLACED_TABLE where {{column}}=1;", JarvisUtil.resolvePlaceholders(query, parameters));
```
Comment (csabakassai) [inline src/test/java/ai/aliz/jarvis/util/TestJarvisUtil.java:34]: throw an exception if any placeholder won't get resolved

**Review by fpeter8 — CHANGES_REQUESTED** [review]
> Reviewed 3 of 21 files at r1, 8 of 22 files at r4, 19 of 23 files at r5, 1 of 1 files at r7, 1 of 1 files at r9, 7 of 7 files at r10.
> *[Reviewable](https://reviewable.io/reviews/aliz-ai/jarvis/17)* status: all files reviewed, 13 unresolved discussions (waiting on @csabakassai and @nora-ambroz)

___
*[src/main/java/ai/aliz/jarvis/config/AppConfig.java, line 13 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPM_F6B21rNxMdQBwX:-MRPM_F72_8shE5igyWV:bbntj4e) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/config/AppConfig.java#L13)):*
> ```Java
> @Getter
> @Setter
> public class AppConfig {
> ```

`@Data`?
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 63 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMkkSE_6HTS0oNVFb:-MRPMkkSE_6HTS0oNVFc:b-dlmkw2) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L63)):*
> ```Java
>     public void executeStatement(String query, Context context) {
>         TableResult queryResult = executeQueryAndGetResult(query, context);
>         log.debug("Query result {}", queryResult.iterateAll());
> ```

I don't think we should log all lines returned from BigQuery
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 70 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMuvUCN8wefCtEES9:-MRPMuvUCN8wefCtEESA:bhn8nuz) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L70)):*
> ```Java
>         TableResult queryResult = executeQueryAndGetResult(query, context);
>         ArrayNode result = bigQueryResultToJsonArrayNode(queryResult);
>         return result.toString();
> ```

what is the `toString()` of the `TableResult` class?
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 77 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPN265CIdZiHWEToyJ:-MRPN265CIdZiHWEToyK:b-as716h) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L77)):*
> ```Java
>         List<String> deletes = Lists.newArrayList();
>         List<String> inserts = Lists.newArrayList();
>         Arrays.stream(query.split(";"))
> ```

this is a very crude/naive implementation; please add a TODO/FIXME and a new ticket to do this properly.
ways this can be broken:
- `;` in the middle of a string constant (`SELECT ';' AS col1;`)
- `;` in comments (
```
SELECT * -- we use star here; deal with it
FROM table
```
)

another problem with this is that you parallelize the script when there is no proof that they are independent of each other
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 104 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPNt3t-vHR-liRnaMx:-MRPNt3t-vHR-liRnaMy:b-81bugd) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L104)):*
> ```Java
>     public Long getTableLastModifiedAt(Context context, String project, String dataset, String table) {
>         BigQuery bigQuery = getBigQueryClient(context);
>         log.info("Getting last modified at for table: {}.{}.{}", project, dataset, table);
> ```

I'd move this to debug; this floods the test logs unnecessarily
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 126 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPO3zwBk4jVDefxzXB:-MRPO3zx-dF9MtLrgms0:bgrbw85) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L126)):*
> ```Java
>         BigQuery bigQuery = getBigQueryClient(context);
>         try {
>             log.info("Executing {}", query);
> ```

log the completed query here too

also I like to wrap the placeholder between quotes, that way I know exactly where the query ended (`Executing '{}'`)
___
*[src/main/java/ai/aliz/jarvis/db/MxSQLQueryExecutor.java, line 39 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPOvjX3DAa7YSCefd2:-MRPOvjX3DAa7YSCefd3:b-samonv) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/MxSQLQueryExecutor.java#L39)):*
> ```Java
>             "jdbc:sqlserver://{{host}}:{{port}};databaseName={{database}};user={{user}};password={{password}}";
>     
>     private static final String My_CONNECTION_STRING_PATTERN =
> ```

lowercase y
___
*[src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java, line 86 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPPWF57TRmQ5LPNQti:-MRPPWF57TRmQ5LPNQtj:b-7f7hu6) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java#L86)):*
> ```Java
>         BigQuery bigQuery = BigQueryOptions.newBuilder().setProjectId(project).build().getService();
>         TableDataWriteChannel writer = bigQuery.writer(writeChannelConfiguration);
>         try (OutputStream stream = Channels.newOutputStream(writer)) {
> ```

does this propagate the `.close()` method to the writer?
___
*[src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java, line 96 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPPaKC91Me1WiWPrhG:-MRPPaKC91Me1WiWPrhH:b2fu745) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java#L96)):*
> ```Java
>         Job job = writer.getJob();
>         try {
>             job = job.waitFor();
> ```

you should add a timeout here (maybe copy from Merkur `dk.dagrofa.merkur.bigquery.BigQueryUtil#waitForJobsToComplete(java.util.stream.Stream<com.google.cloud.bigquery.Job>, java.time.Instant)`)
___
*[src/main/java/ai/aliz/jarvis/service/shared/platform/SFTPService.java, line 86 at r9](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMS8x7btHIEWtrHhu-r9-86:-MRPQ7DAEkB9NTy1oyKL:bynipiv) ([raw file](https://github.com/aliz-ai/jarvis/blob/d6a54e9742bb32850fd22b9216aa2bb44c879dfd/src/main/java/ai/aliz/jarvis/service/shared/platform/SFTPService.java#L86)):*
<details><summary><i>Previously, nora-ambroz wrote…</i></summary><blockquote>

What should I do here? 
</blockquote></details>

I guess nothing, but @csabakassai

**Review by nora-ambroz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aliz-ai/jarvis/17)* status: 32 of 40 files reviewed, 10 unresolved discussions (waiting on @csabakassai and @fpeter8)

___
*[README.MD, line 157 at r9](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMS8zpDF0SwbmqKGi-r9-157:-MRzSFoT0OL9LBTYtfka:b-896fig) ([raw file](https://github.com/aliz-ai/jarvis/blob/d6a54e9742bb32850fd22b9216aa2bb44c879dfd/README.MD#L157)):*
<details><summary><i>Previously, csabakassai (Kassai Csaba) wrote…</i></summary><blockquote>

I think the manual setup is implemted anymore, so just please set the default value to autodetect to true and this way it does not need to be explicitly stated in the testSuite.json file. The manual configuration is not needed anymore. 
Also please document the config for different init actions separately
1. BQLoad
2. SQLExec
3. SFTPLoad
And how is the init action type determined based on the context and the file extensions etc...
</blockquote></details>

Done?
___
*[README.MD, line 145 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMS8zpDF0SwbmqKGi-r10-145:-MRymNlKEoR2QFlQneBR:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/README.MD#L145)):*
<details><summary><i>Previously, csabakassai (Kassai Csaba) wrote…</i></summary><blockquote>

In the code you call this 'Step'. I think that is better name, please use also stepo here in the docs 
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/config/AppConfig.java, line 13 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPM_F6B21rNxMdQBwX:-MRzSWJ3DhiNvdhtTdFu:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/config/AppConfig.java#L13)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

`@Data`?
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 63 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMkkSE_6HTS0oNVFb:-MRzTyrk2DEbE-aAXF1I:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L63)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

I don't think we should log all lines returned from BigQuery
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 70 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMuvUCN8wefCtEES9:-MRzUllQ3jYt-4zkpR-N:boowne2) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L70)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

what is the `toString()` of the `TableResult` class?
</blockquote></details>

```
MoreObjects.toStringHelper(this)
        .add("rows", getValues())
        .add("schema", schema)
        .add("totalRows", totalRows)
        .add("cursor", getNextPageToken())
        .toString();
```
Do you think this is okay to use here?
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 77 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPN265CIdZiHWEToyJ:-MRzrBOHD0uhAn0yQhKE:b-ryj4pt) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L77)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

this is a very crude/naive implementation; please add a TODO/FIXME and a new ticket to do this properly.
ways this can be broken:
- `;` in the middle of a string constant (`SELECT ';' AS col1;`)
- `;` in comments (
```
SELECT * -- we use star here; deal with it
FROM table
```
)

another problem with this is that you parallelize the script when there is no proof that they are independent of each other
</blockquote></details>

Done. Do you have a suggestion how to handle this instead? It can maybe fit into the time budget.
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 80 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMS8wyfbGjNWCag8X-r10-80:-MRzV5SW616uc_08eKYl:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L80)):*
<details><summary><i>Previously, nora-ambroz wrote…</i></summary><blockquote>

Should not change to upper case.
Fixed on init-action-integration-tests branch.
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 104 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPNt3t-vHR-liRnaMx:-MRzVN9sDg8tSQ3d8nz3:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L104)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

I'd move this to debug; this floods the test logs unnecessarily
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 126 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPO3zwBk4jVDefxzXB:-MRzVwJEDIvn9OX_4wP9:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L126)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

log the completed query here too

also I like to wrap the placeholder between quotes, that way I know exactly where the query ended (`Executing '{}'`)
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/db/MxSQLQueryExecutor.java, line 39 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPOvjX3DAa7YSCefd2:-MRzWCSLAiOudpWnn-F2:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/MxSQLQueryExecutor.java#L39)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

lowercase y
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java, line 86 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPPWF57TRmQ5LPNQti:-MRzY-39D2s8BO19fbMa:b-ga4w6v) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java#L86)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

does this propagate the `.close()` method to the writer?
</blockquote></details>

According to the docs 'Closing the stream will in turn cause the channel to be closed.' So the other way around? What changes does this require?
___
*[src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java, line 96 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPPaKC91Me1WiWPrhG:-MRzdjXz6g3NqZ6AY35P:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java#L96)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

you should add a timeout here (maybe copy from Merkur `dk.dagrofa.merkur.bigquery.BigQueryUtil#waitForJobsToComplete(java.util.stream.Stream<com.google.cloud.bigquery.Job>, java.time.Instant)`)
</blockquote></details>

Done.
___
*[src/test/java/ai/aliz/jarvis/util/TestJarvisUtil.java, line 34 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMS8ynG8m3KobEXS6-r10-34:-MRznweN3aezns8AObCA:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/test/java/ai/aliz/jarvis/util/TestJarvisUtil.java#L34)):*
<details><summary><i>Previously, csabakassai (Kassai Csaba) wrote…</i></summary><blockquote>

throw an exception if any placeholder won't get resolved
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/service/shared/platform/SFTPService.java, line 86 at r9](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPMS8x7btHIEWtrHhu-r9-86:-MRzdnx5DS3flhjvnSZ5:b-f7e8o6) ([raw file](https://github.com/aliz-ai/jarvis/blob/d6a54e9742bb32850fd22b9216aa2bb44c879dfd/src/main/java/ai/aliz/jarvis/service/shared/platform/SFTPService.java#L86)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

I guess nothing, but @csabakassai 
</blockquote></details>

@csabakassai  ping?

**Review by fpeter8 — CHANGES_REQUESTED** [review]
> Reviewed 18 of 21 files at r1, 5 of 5 files at r2, 14 of 22 files at r4, 8 of 8 files at r11.
> Dismissed @csabakassai from a discussion.
> *[Reviewable](https://reviewable.io/reviews/aliz-ai/jarvis/17)* status: all files reviewed, 3 unresolved discussions (waiting on @csabakassai and @nora-ambroz)

___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 77 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPN265CIdZiHWEToyJ:-MSD5ozB8a14c82H-quk:bzb0yn7) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L77)):*
<details><summary><i>Previously, nora-ambroz wrote…</i></summary><blockquote>

Done. Do you have a suggestion how to handle this instead? It can maybe fit into the time budget.
</blockquote></details>

copy from another project:

```java
    /**
     * Splits a SQL script into its statements. Take this example: <br>
     * <pre>
     *     INSERT INTO a (ID, VALUE) VALUES (1, 'A');
     *     SELECT * FROM a;
     * </pre>
     * The method would return two statements: <code>INSERT INTO a (ID, VALUE) VALUES (1, 'A')</code> and <code>SELECT * FROM a</code>.
     * <p>
     * This method also strips away all comments from the script, leaving only the actual SQL behind.
     *
     * @param script The SQL script as a string.
     * @return a list of SQL statements.
     */
    public List<String> splitScriptIntoStatements(@NonNull String script) {
        List<String> statements = new ArrayList<>();
        if (script.length() == 0) {
            return statements;
        }

        StringBuilder sb = new StringBuilder();

        // whether we're in the middle of a ', " or `
        boolean inConstant = false;
        // the character bordering the constant
        char constantStart = 0;

        // if we have started a -- comment in code
        boolean dashDashStarted = false;
        // if we activated single line comment
        boolean inSingleLineComment = false;

        // whether we just started /* or */
        boolean slashStarStarted = false;

        // if we activated multi line comment
        boolean inMultiLineComment = false;

        for (char token : script.toCharArray()) {
            if (inConstant) {
                // if we're in a constant, output no matter what; at the end, clear flag
                sb.append(token);
                if (token == constantStart) {
                    inConstant = false;
                    constantStart = 0;
                }
            } else if (inSingleLineComment) {
                // if we are at the end of the line, clear flag; otherwise don't output anything
                if (token == '\n') {
                    sb.append(token);
                    inSingleLineComment = false;
                }
            } else if (inMultiLineComment) {
                if (slashStarStarted) {
                    // clear this flag no matter what
                    slashStarStarted = false;
                    if (token == '/') {
                        // detected '*/' -> stop comment mode
                        inMultiLineComment = false;
                        sb.append("\n");
                    }
                }
                if (token == '*') {
                    // might be a start of '*/'
                    slashStarStarted = true;
                }
            } else if (dashDashStarted) {
                // clear flag no matter what
                dashDashStarted = false;
                if (token == '-') {
                    // '--' detected -> comment
                    inSingleLineComment = true;
                } else {
                    // not '--' -> have to output the previous '-' as well as this token
                    sb.append('-').append(token);
                }
            } else if (slashStarStarted) {
                // clear flag no matter what
                slashStarStarted = false;
                if (token == '*') {
                    // '/*' detected; multi line comment
                    inMultiLineComment = true;
                } else {
                    // not '/*' -> have to output '/' and current token
                    sb.append('/').append(token);
                }
            } else {
                // at this point we can assume all flags are false; any flag should be handled prior to this branch
                switch (token) {
                    case '-':
                        dashDashStarted = true;
                        break;
                    case '#':
                        inSingleLineComment = true;
                        break;
                    case '/':
                        slashStarStarted = true;
                        break;
                    case '\'':
                    case '"':
                    case '`':
                        inConstant = true;
                        constantStart = token;
                        sb.append(token);
                        break;
                    case ';':
                        // save this statement, and start a new one
                        String statement = sb.toString().trim();
                        if (statement.length() > 0) {
                            statements.add(statement);
                        }
                        sb.setLength(0);
                        break;
                    default:
                        sb.append(token);
                        break;
                }
            }
        }

        String statement = sb.toString().trim();
        if (statement.length() > 0) {
            statements.add(statement);
        }

        return statements;
    }
```
___
*[src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java, line 86 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPPWF57TRmQ5LPNQti:-MSD6As73Am9sHKctNmO:bhdbwn4) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/service/init/initiator/BQLoadInitiator.java#L86)):*
<details><summary><i>Previously, nora-ambroz wrote…</i></summary><blockquote>

According to the docs 'Closing the stream will in turn cause the channel to be closed.' So the other way around? What changes does this require?
</blockquote></details>

in that case nothing
___
*[src/main/java/ai/aliz/jarvis/util/JarvisUtil.java, line 72 at r11](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MSD6H1xBZZrCC7UQUba:-MSD6H1xBZZrCC7UQUbb:b-bsac5o) ([raw file](https://github.com/aliz-ai/jarvis/blob/12cd6769be09e4cce0678eef2c61e42c5332ee5b/src/main/java/ai/aliz/jarvis/util/JarvisUtil.java#L72)):*
> ```Java
>             result = result.replace("{{" + kv.getKey() + "}}", kv.getValue());
>         }
>         if (result.matches("^.*\\{\\{.*?}}.*$")) {
> ```

not super important, but you should probably have a `private static final Pattern = ...` field; this re-parses the entire regex every time you call this method

**Review by fpeter8 — COMMENTED** [review]
> Dismissed @csabakassai from a discussion.
> *[Reviewable](https://reviewable.io/reviews/aliz-ai/jarvis/17)* status: all files reviewed, 2 unresolved discussions (waiting on @csabakassai and @nora-ambroz)

**Review by nora-ambroz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aliz-ai/jarvis/17)* status: 38 of 40 files reviewed, 1 unresolved discussion (waiting on @fpeter8)

___
*[src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java, line 77 at r10](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MRPN265CIdZiHWEToyJ:-MSTW8on4NiZiapC5pbr:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/d57cf3b153f59d75a5f6a2f6f0f1be399d4c4f2e/src/main/java/ai/aliz/jarvis/db/BigQueryExecutor.java#L77)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

copy from another project:

```java
    /**
     * Splits a SQL script into its statements. Take this example: <br>
     * <pre>
     *     INSERT INTO a (ID, VALUE) VALUES (1, 'A');
     *     SELECT * FROM a;
     * </pre>
     * The method would return two statements: <code>INSERT INTO a (ID, VALUE) VALUES (1, 'A')</code> and <code>SELECT * FROM a</code>.
     * <p>
     * This method also strips away all comments from the script, leaving only the actual SQL behind.
     *
     * @param script The SQL script as a string.
     * @return a list of SQL statements.
     */
    public List<String> splitScriptIntoStatements(@NonNull String script) {
        List<String> statements = new ArrayList<>();
        if (script.length() == 0) {
            return statements;
        }

        StringBuilder sb = new StringBuilder();

        // whether we're in the middle of a ', " or `
        boolean inConstant = false;
        // the character bordering the constant
        char constantStart = 0;

        // if we have started a -- comment in code
        boolean dashDashStarted = false;
        // if we activated single line comment
        boolean inSingleLineComment = false;

        // whether we just started /* or */
        boolean slashStarStarted = false;

        // if we activated multi line comment
        boolean inMultiLineComment = false;

        for (char token : script.toCharArray()) {
            if (inConstant) {
                // if we're in a constant, output no matter what; at the end, clear flag
                sb.append(token);
                if (token == constantStart) {
                    inConstant = false;
                    constantStart = 0;
                }
            } else if (inSingleLineComment) {
                // if we are at the end of the line, clear flag; otherwise don't output anything
                if (token == '\n') {
                    sb.append(token);
                    inSingleLineComment = false;
                }
            } else if (inMultiLineComment) {
                if (slashStarStarted) {
                    // clear this flag no matter what
                    slashStarStarted = false;
                    if (token == '/') {
                        // detected '*/' -> stop comment mode
                        inMultiLineComment = false;
                        sb.append("\n");
                    }
                }
                if (token == '*') {
                    // might be a start of '*/'
                    slashStarStarted = true;
                }
            } else if (dashDashStarted) {
                // clear flag no matter what
                dashDashStarted = false;
                if (token == '-') {
                    // '--' detected -> comment
                    inSingleLineComment = true;
                } else {
                    // not '--' -> have to output the previous '-' as well as this token
                    sb.append('-').append(token);
                }
            } else if (slashStarStarted) {
                // clear flag no matter what
                slashStarStarted = false;
                if (token == '*') {
                    // '/*' detected; multi line comment
                    inMultiLineComment = true;
                } else {
                    // not '/*' -> have to output '/' and current token
                    sb.append('/').append(token);
                }
            } else {
                // at this point we can assume all flags are false; any flag should be handled prior to this branch
                switch (token) {
                    case '-':
                        dashDashStarted = true;
                        break;
                    case '#':
                        inSingleLineComment = true;
                        break;
                    case '/':
                        slashStarStarted = true;
                        break;
                    case '\'':
                    case '"':
                    case '`':
                        inConstant = true;
                        constantStart = token;
                        sb.append(token);
                        break;
                    case ';':
                        // save this statement, and start a new one
                        String statement = sb.toString().trim();
                        if (statement.length() > 0) {
                            statements.add(statement);
                        }
                        sb.setLength(0);
                        break;
                    default:
                        sb.append(token);
                        break;
                }
            }
        }

        String statement = sb.toString().trim();
        if (statement.length() > 0) {
            statements.add(statement);
        }

        return statements;
    }
```
</blockquote></details>

Done.
___
*[src/main/java/ai/aliz/jarvis/util/JarvisUtil.java, line 72 at r11](https://reviewable.io/reviews/aliz-ai/jarvis/17#-MSD6H1xBZZrCC7UQUba:-MSTPeZc8Udi9B9dbnPq:b-896fix) ([raw file](https://github.com/aliz-ai/jarvis/blob/12cd6769be09e4cce0678eef2c61e42c5332ee5b/src/main/java/ai/aliz/jarvis/util/JarvisUtil.java#L72)):*
<details><summary><i>Previously, fpeter8 (Peter Farkas) wrote…</i></summary><blockquote>

not super important, but you should probably have a `private static final Pattern = ...` field; this re-parses the entire regex every time you call this method
</blockquote></details>

Done.

**Review by fpeter8 — APPROVED** [review]
> Reviewed 2 of 2 files at r12, 1 of 1 files at r13.
> *[Reviewable](https://reviewable.io/reviews/aliz-ai/jarvis/17)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nora-ambroz)

**Pull request merged**